### ISSUE TITLE
WebView2 setup fails to cleanup old version if previous directories are empty

### ISSUE THREAD

**Issue by amahnk** [MicrosoftEdge/WebView2Feedback#1946]
The Edge WebView2 setup cleanup routine will hang and fail to remove previous version directories in _Program Files X86\Microsoft\EdgeWebView\Application_ if the previous version directories do not contain the file **msedgewebview2.exe** . 

Example:
`"C:\Program Files (x86)\Microsoft\EdgeWebView\Application\95.0.1020.53\Installer\setup.exe" --msedgewebview --delete-old-versions --system-level --verbose-logging --on-logon`

Because this occurs during logon by RunOnce wrapper, the login process will essentially hang with a black screen since the Explorer shell will not load until RunOnce has completed. Additionally, the msedge_installer.log will erroneously report that the file is in use even though it does not exist:

`WARNING:delete_old_versions.cc(167)] Failed to delete old executable on C:\Program Files (x86)\Microsoft\EdgeWebView\Application\95.0.1020.30\msedgewebview2.exe: The system cannot find the file specified. (0x2)`

This can be reproduced as easily as creating a blank directory in _Program Files X86\Microsoft\EdgeWebView\Application_ that matches the version numbering scheme (e.g. 99.9.9999.99). Simply adding a blank file with the filename **msedgewebview2.exe** to the same directory then allows the setup process to complete successfully.

Adam

**Comment by champnic** [comment]
Thanks for the bug report @amahnk! I've forwarded it to our deployment team.

**Comment by amahnk** [comment]
After further testing in my environment, I discovered that the the state of HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\AsyncRunOnce is what is actually causing the black screen during login. If AsyncRunOnce is set to 0, a black screen will appear since Windows is being instructed to wait until RunOnce completes. Setting AsyncRunOnce to 1 prevents this. However the bug still exists in that the setup exec is using flawed logic when deleting orphaned directories and should be fixed -- I have been periodically running a small script through an endpoint management platform to delete the old directories as a workaround.

Adam

**Comment by agoyal03** [comment]
is there any update on this ,this is creating bad experience for thousands of users ,where our App Install itself is getting stuck ,because of Edge installer going into infinite loop .

**Comment by jadelise** [comment]
duplicated this error with latest Citrix Workspace Installer, Citrix VDA 7.15 and Citrix VDA 1912 CU4. Hangs forever and the work arounds were not working for me or were to hard to code in my installation wrapper.

I found my own work around.  Kill setup.exe, the installation completes normally. I should be able to code that very easily, watch for setup.exe to appear, get the processes command line to match a pattern on so I not killing the wrong setup.exe  and the kill the process.  That's my plan.  Edge does not seem to be effected by the termination of the process.

**Comment by champnic** [comment]
@agoyal03 @jadelise Can you confirm that you are hitting the same issue as @amahnk first reported? Do the workarounds they describe fix the issue for you? Thanks!

**Comment by agoyal03** [comment]
@champnic  yes it is same issue as reported and yes workaround is just to go and delete any leftover folder/files  from following location  and it fixes .
C:\Program Files (x86)\Microsoft\EdgeWebView\Application
And as mentioned above it could be easily reproduced by creating a blank file in above location . This has been reported by lots of end users for our application  and it's not possible for each end user to individually apply workarounds.

**Comment by champnic** [comment]
Thanks for confirming @agoyal03! Our installer team is working on a fix now.
